pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go

```diff
@@ -83,7 +83,11 @@ func newResourceModel(logger klog.Logger, resourceSliceLister resourcev1alpha2li
 			if model[structured.NodeName] == nil {
 				model[structured.NodeName] = make(map[string]ResourceModels)
 			}
-			resource := model[structured.NodeName][handle.DriverName]
+			driverName := handle.DriverName
+			if driverName == "" {
+				driverName = claim.Status.DriverName
+			}
+			resource := model[structured.NodeName][driverName]
```

Comment:
I remember this coming up in the past, but I don't remember. When might `handle.DriverName == ""`?

Reply:
It's simply how we defined out API:
```
// ResourceHandle holds opaque resource data for processing by a specific kubelet plugin.
type ResourceHandle struct {
	// DriverName specifies the name of the resource driver whose kubelet
	// plugin should be invoked to process this ResourceHandle's data once it
	// lands on a node. This may differ from the DriverName set in
	// ResourceClaimStatus this ResourceHandle is embedded in.
	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
```

Perhaps we should have made it required to avoid if checks like this one here. Instead, we made it optional to allow avoiding redundant values in the claim status.

Reply:
Actually... validation requires it:

https://github.com/kubernetes/kubernetes/blob/634fc1b4836b3a500e0d715d71633ff67690526a/pkg/apis/resource/validation/validation.go#L201 

That'll complain if the name is empty.

All questions about whether we could change validation aside (implies API break), I'm leaning towards keeping the validation as-is and fixing the API definition. This would make this PR unnecessary for 1.30 because only a small typo fix remains and also avoids this potential pitfall in the future.

Reply:
+1 on fixing API definition.

Reply:
I remember now why this is necessary. In "traditional" DRA, the driver's controller is responsible for populating the ResourceHandle. If the driver doesn't actually use the ResourceHandle to communicate information to the kubelet plugin (as is the case with the NVIDIA driver), then it shouldn't have to instantiate one just to set this DriverName field in it.

Reply:
It's the reason we have this in the kubelet:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager.go#L135-L140

Reply:
>  If the driver doesn't actually use the ResourceHandle to communicate information to the kubelet plugin (as is the case with the NVIDIA driver), then it shouldn't have to instantiate one just to set this DriverName field in it.

Then there is no `ResourceHandle` to validate. That's unrelated to whether `DriverName` must be set when there is a `ResourceHandle`.

> It's the reason we have this in the kubelet:
> https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager.go#L135-L140

Here I don't quite follow. That code only gets called if there is a `ResourceHandle`. Or is that using a fake `ResourceHandle` in `claimInfo.ResourceHandles` that didn't actually from from the claim status? If that is so, then why not copy in the right `DriverName` when faking the handle?

Reply:
See https://github.com/kubernetes/kubernetes/pull/124075 for the API fix.

/close